app/src/main/java/io/gnosis/kouban/push/PushServiceRepository.kt

```diff
@@ -0,0 +1,202 @@
+package io.gnosis.kouban.push
+
+import android.content.Context
+import androidx.navigation.NavDeepLinkBuilder
+import com.google.android.gms.tasks.OnCompleteListener
+import com.google.firebase.iid.FirebaseInstanceId
+import io.gnosis.kouban.R
+import io.gnosis.kouban.core.managers.SafeAddressManager
+import io.gnosis.kouban.data.backend.PushServiceApi
+import io.gnosis.kouban.helpers.LocalNotificationManager
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.runBlocking
+import pm.gnosis.model.Solidity
+import pm.gnosis.utils.asEthereumAddressString
+
+class PushServiceRepository(
+    private val context: Context,
+    private val localNotificationManager: LocalNotificationManager,
+    private val pushServiceApi: PushServiceApi,
+    private val prefs: PushPrefs,
+    private val safeAddressManager: SafeAddressManager
+) {
+
+    fun checkRegistration() {
+
+        if (!prefs.isDeviceRegistered) {
+            FirebaseInstanceId.getInstance().instanceId.addOnCompleteListener(OnCompleteListener { task ->
+                if (!task.isSuccessful) {
+                    return@OnCompleteListener
+                }
+                // Get new Instance ID token
+                val token = task.result?.token
+                token?.let {
+                    registerDevice(it)
+                }
+            })
+        } else {
+            if (!prefs.isSafeRegistered) {
+                runBlocking(Dispatchers.IO) {
```

Comment:
I do agree with @jpalvarezl here. We should try to register the Safe if the device was registered. Right now it would only do that on the second call to this method